rfcs/0000-multiple-kuadrant-instances.md

```diff
@@ -0,0 +1,152 @@
+
+### The Kuadrant CRD
+
+Currently, the Kuadrant CRD has an empty spec.
+
+```yaml
+apiVersion: kuadrant.io/v1beta1
+kind: Kuadrant
+metadata:
+  name: kuadrant-sample
+spec: {}
+```
+
+According to the definition above of the kuadrant instance,
+the proposed new Kuadrant CRD would add a label __selector__ to specify which gateways that instance would manage.
+
+```yaml
+apiVersion: kuadrant.io/v1beta1
+kind: Kuadrant
+metadata:
+  name: kuadrant-a
+spec:
+  gatewaysSelector:
+    matchLabels:
+      app: kuadrant-a
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+### Wiring Kuadrant policies with Kuadrant instances
+Technically, the Kuadrant policies do not belong to any Kuadrant instance. At any moment of time, one policy can switch the targeted network resource specified in the `spec` from one gateway to another. Directly or indirectly via the HTTPRoute. The target references are dynamic by nature, so is the list of gateways to which kuadrant policies should apply.
+Thus, the Kuadrant's control plane needs a procedure to associate a policy with **one** kuadrant instance at any time. When the control plane knows which kuadrant instance is affected, the policy rules can be used to configure the Limitador and Authorino instances belonging to that kuadrant's instance. Since the associated kuadrant instance of a policy is dynamic by nature, this procedure must be executed on  every event related to the policy.
+
+When the policy's `targetRef` targets a Gateway, there is a direct reference to the gateway.
+
+When the policy's `targetRef` targets an HTTPRoute, Kuadrant will follow the [`parentRef`](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1beta1.CommonRouteSpec) attribute which should be a direct reference to the gateway or gateways.
+
+Given a gateway, Kuadrant needs to find out which Kuadrant's instance is managing that specific gateway. By design, Kuadrant knows it is only one. There are at least two options to implement that mapping:
+* Read all Kuadrant CR objects and the first one that matches label selector.
+  * This approach works as long as the control plane ensures that each gateway is matched by only one kuadrant gateway selector. The control plane must reject any new kuadrant instance matching a gateway already "taken" by other kuadrant instance.
+* Add annotation in the gateway with a value of the Name/Namespace of the Kuadrant CR.
+  * This approach is commonly used. Requires annotation management.
```

Comment:
We could define the list of gateways that meant to be managed in the Kuadrant CR.

Reply:
I'd suggest a list of `GatewayClasses` instead of `Gateways`.

Users can always "split" one `GatewayClass` into two or more, handled by the same controller or not, to highlight the fact that different configurations apply to each class. In fact, this is encouraged by [GW-API](https://gateway-api.sigs.k8s.io/api-types/gatewayclass/):

> We expect that one or more GatewayClasses will be created by the infrastructure provider for the user. It allows decoupling of which mechanism (e.g. controller) implements the Gateways from the user. For instance, an infrastructure provider may create two GatewayClasses named internet and private to reflect Gateways that define Internet-facing vs private, internal applications.

 Some discussion about this also here: https://github.com/Kuadrant/architecture/pull/7#issuecomment-1397383619

Reply:
Kuadrant manages gateway instances, not gateway classes. Kuadrant instance A may manage gateway X and instance B may mananage instance Y. Both X and Y gateways may be share the same gateway class.

Reply:
I guess what I'm saying is that Kuadrant can use the GatewayClass to know which Gateways to manage.

In your provided example, 

> Kuadrant instance A may manage gateway X and instance B may mananage instance Y

1. You can have both instances of Kuadrant listing the one and only GatewayClass; or
2. You can have 2 GatewayClasses, one for each gateway, one for each instance of Kuadrant.